server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,10 +233,158 @@ private void performOnReplicas(
-                performOnReplica(shard, replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, pendingReplicationActions);
+        for (final ReplicationModeAwareShardRouting shardRouting : replicationGroup.getReplicationTargets()) {
+            replicationProxy.performOnReplica(
+                shardRouting,
+                primaryRouting,
+                replicaRequest,
+                globalCheckpoint,
+                maxSeqNoOfUpdatesOrDeletes,
+                pendingReplicationActions
+            );
+        }
+    }
+
+    /**
+     * Factory that return the {@link ReplicationProxy} instance basis the {@link ReplicationModeAwareOverrideProxy}.
+     *
+     * @opensearch.internal
+     */
+    private class ReplicationProxyFactory {
+        ReplicationProxy create(final Optional<ReplicationOverridePolicy> overridePolicy) {
+            if (overridePolicy.isEmpty()) {
+                return new FanoutReplicationProxy();
+            } else {
+                return new ReplicationModeAwareOverrideProxy(overridePolicy.get());
+            }
+        }
+    }
+
+    /**
+     * Used for performing any replication operation on replicas. Depending on the implementation, the replication call
+     * can fanout or stops here.
+     *
+     * @opensearch.internal
+     */
+    private abstract class ReplicationProxy {
+
+        /**
+         * Depending on the actual implementation and the passed {@link ReplicationModeAwareShardRouting}, the replication
+         * mode is determined using which the replication request is performed on the replica or not.
+         *
+         * @param shardRouting replication mode aware ShardRouting
+         * @param primaryRouting primary ShardRouting
+         * @param replicaRequest replication request
+         * @param globalCheckpoint current global checkpoint on primary
+         * @param maxSeqNoOfUpdatesOrDeletes maxSeqNoOfUpdatesOrDeletes
+         * @param pendingReplicationActions pendingReplicationActions
+         */
+        private void performOnReplica(
+            final ReplicationModeAwareShardRouting shardRouting,
+            final ShardRouting primaryRouting,
+            final ReplicaRequest replicaRequest,
+            final long globalCheckpoint,
+            final long maxSeqNoOfUpdatesOrDeletes,
+            final PendingReplicationActions pendingReplicationActions
+        ) {
+            ReplicationMode replicationMode = determineReplicationMode(shardRouting, primaryRouting);
+            // If the replication modes are 1. Logical replication or 2. Primary term validation, we let the call get performed on the
+            // replica shard.
+            if (replicationMode == ReplicationMode.LOGICAL_REPLICATION || replicationMode == ReplicationMode.PRIMARY_TERM_VALIDATION) {
+                ReplicationOperation.this.performOnReplica(
+                    shardRouting.getShardRouting(),
+                    replicaRequest,
```

Comment:
do we send to the same `replicaRequest` in case of `LOGICAL_REPLICATION` and `PRIMARY_TERM_VALIDATION` ?

Reply:
> do we send to the same replicaRequest in case of LOGICAL_REPLICATION and PRIMARY_TERM_VALIDATION ?

Yes, for full replication and primary term validation both, we send the request to the replica shard. There will be a follow up PR that would bring a thin layer for primary term validation (basically perform primary term valdiation upfront and fail fast if stale).